libeos-payg/boottime-source.c

```diff
@@ -0,0 +1,154 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2018 Endless Mobile, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Alternatively, the contents of this file may be used under the terms of the
+ * GNU Lesser General Public License Version 2.1 or later (the "LGPL"), in
+ * which case the provisions of the LGPL are applicable instead of those above.
+ * If you wish to allow use of your version of this file only under the terms
+ * of the LGPL, and not to allow others to use your version of this file under
+ * the terms of the MPL, indicate your decision by deleting the provisions
+ * above and replace them with the notice and other provisions required by the
+ * LGPL. If you do not delete the provisions above, a recipient may use your
+ * version of this file under the terms of either the MPL or the LGPL.
+ */
+
+#include "config.h"
+
+#include <errno.h>
+#include <inttypes.h>
+#include <sys/timerfd.h>
+#include <unistd.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <libeos-payg/boottime-source.h>
+
+typedef struct _EpgBoottimeSource {
+  GSource parent;
+
+  int fd;
+  gpointer tag;
+} EpgBoottimeSource;
+
+static gboolean
+epg_boottime_source_check (GSource *source)
+{
+  EpgBoottimeSource *self = (EpgBoottimeSource *) source;
+
+  return g_source_query_unix_fd (source, self->tag) != 0;
+}
+
+static gboolean
+epg_boottime_source_dispatch (GSource    *source,
+                                      GSourceFunc callback,
+                                      gpointer    user_data)
+{
+  EpgBoottimeSource *self = (EpgBoottimeSource *) source;
+  uint64_t n_expirations = 0;
+
+  if (!callback)
+    {
+      g_warning ("Boottime source dispatched without callback. "
+                 "You must call g_source_set_callback().");
+      return G_SOURCE_REMOVE;
+    }
+
+  /* Must read from the FD to reset its ready state. */
+  if (read (self->fd, &n_expirations, sizeof n_expirations) < 0)
+    {
+      g_warning ("read() failed for timerfd: %s",
+                 g_strerror (errno));
+      return G_SOURCE_REMOVE;
+    }
+
+  return callback (user_data);
+}
+
+static void
+epg_boottime_source_finalize (GSource *source)
+{
+  EpgBoottimeSource *self = (EpgBoottimeSource *) source;
+  g_autoptr(GError) error = NULL;
+
+  if (!g_close (self->fd, &error))
+    g_warning ("failed to close timerfd: %s", error->message);
+  self->fd = -1;
+}
+
+static GSourceFuncs epg_boottime_source_funcs = {
```

Comment:
```suggestion
static const GSourceFuncs epg_boottime_source_funcs = {
```

Reply:
With that we have to cast it when using it in `g_source_new` so it doesn't seem worth it (even though `g_source_new` seems to treat it as `const`).

Reply:
It’s worth it because it documents your intent for the structure, and (without checking the assembly) means the compiler can put it in a read-only section of the object file.